diff --git a/vace/models/wan/modules/model.py b/vace/models/wan/modules/model.py
--- a/vace/models/wan/modules/model.py
+++ b/vace/models/wan/modules/model.py
@@ -34,14 +34,13 @@ class VaceWanAttentionBlock(WanAttentionBlock):
         if self.block_id == 0:
             c = self.before_proj(c) + x
             all_c = []
         else:
-            all_c = list(torch.unbind(c))
+            all_c = list(c)
             c = all_c.pop(-1)
         c = super().forward(c, **kwargs)
         c_skip = self.after_proj(c)
-        all_c += [c_skip, c]
-        c = torch.stack(all_c)
-        return c
+        all_c += [c_skip.cpu(), c]
+        return tuple(all_c)
     
     
 class BaseWanAttentionBlock(WanAttentionBlock):
@@ -63,7 +62,8 @@ class BaseWanAttentionBlock(WanAttentionBlock):
     def forward(self, x, hints, context_scale=1.0, **kwargs):
         x = super().forward(x, **kwargs)
         if self.block_id is not None:
-            x = x + hints[self.block_id] * context_scale
+            hint = hints[self.block_id].to(device=x.device, dtype=x.dtype)
+            x = x + hint * context_scale
         return x
     
     
@@ -138,7 +138,7 @@ class VaceWanModel(WanModel):
 
         for block in self.vace_blocks:
             c = block(c, **new_kwargs)
-        hints = torch.unbind(c)[:-1]
+        hints = c[:-1]
         return hints
 
     def forward(
diff --git a/vace/models/wan/distributed/xdit_context_parallel.py b/vace/models/wan/distributed/xdit_context_parallel.py
--- a/vace/models/wan/distributed/xdit_context_parallel.py
+++ b/vace/models/wan/distributed/xdit_context_parallel.py
@@ -90,7 +90,7 @@ def usp_dit_forward_vace(
 
     for block in self.vace_blocks:
         c = block(c, **new_kwargs)
-    hints = torch.unbind(c)[:-1]
+    hints = c[:-1]
     return hints
 
 
